### ISSUE TITLE
Pattern matching, take 2

### ISSUE THREAD

**Issue by wcjohnson** [wcjohnson/lightscript-devkit#9]
Comments and discussion:
https://github.com/wcjohnson/babylon-lightscript/commit/a19aa8b6be5bce4778ae60f0b7ace0e8a8db7097#commitcomment-22356248

https://github.com/lightscript/babylon-lightscript/pull/17

https://gitter.im/lightscript/Lobby

General syntax outline:
```js
match x:
  | unboundTest with pattern if boundTest:
  | unboundTest with pattern:
  | unboundTest as rename:
  | unboundTest:
  | pattern if boundTest:
  | pattern:
  | else as rename:
  | else:
```
- `with` patterns are assertive; generate code to test that the discriminant matches the destructure
- Disallow `PlaceholderExpression`s in `boundTest`.

Modifications:
- Allow non-assertive destructuring with `as` (`parseBindingAtom`)
- Allow `...` (without argument) in array destructuring so as not to needlessly generate rest arrays

Pitfalls:
- Grammatical ambiguity in `unboundTest` between `ArrayPattern` and expressions beginning with an `ArrayExpression`. Possibly requires speculative parsing. E.g. `[a, b, c]~indexOf(d)` should be parsed as an `unboundTest` and not a `pattern`.

**Comment by wcjohnson** [comment]
Further discussion resulted in this simpler syntax:
```js
match x:
  | unboundTest with pattern:
  | unboundTest:
  | pattern:
  | else:
```

Destructuring at the match case level is assertive. Non-assertive destructuring can be done within the blocks:
```
match x:
  | ~isWidget():
    { color, price } = x
```

**Comment by wcjohnson** [comment]
Implemented in 2.0.0-alpha.1
